src/build/__init__.py

```diff
@@ -156,14 +158,25 @@ def __init__(self, srcdir, python_executable=sys.executable):
 
         self._build_system = build_system
         self._backend = self._build_system['build-backend']
-
+        self.script_dir = os.path.dirname(python_executable) if scripts_dir is None else scripts_dir
```

Comment:
Public attributes should be annotated:

```suggestion
        self.script_dir = os.path.dirname(python_executable) if scripts_dir is None else scripts_dir  # type: Optional[Union[bytes, Text]]
```

Reply:
But it's deduced automatically by mypy.

Reply:
The public interface should not be inferred and some type checkers like Pyright do not perform inference on `py.typed` packages at all.

Reply:
I'm tempted to call this a pyright bug. Isn't Python typing meant to be gradual as per https://www.python.org/dev/peps/pep-0484/#abstract (and https://www.python.org/dev/peps/pep-0483/#summary-of-gradual-typing)?

Reply:
If this would hold I'd expect mypy to warn if strict mode is enabled, but it does not 🤔

Reply:
Gradual typing does not refer to inference - it refers to the fact that typing is opt-in.  If you run mypy in strict mode, gradual typing ceases to be an option.  It's not a bug in Pyright that public attributes must be annotated - it's a conscious design decision on their part.  That mypy doesn't error for untyped public attributes is very likely an omission.

Reply:
I see that you've changed this to a read-only property which is annotated, collapsing.